File: sdks/typescript/src/skill.ts
/**
 * Generate SKILL.md files from ACLI command trees.
 *
 * Emits a file conforming to the agentskills.io open standard
 * (https://agentskills.io): YAML frontmatter (`name`, `description`, optional
 * `when_to_use`) followed by the ACLI command reference body.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import type { CommandInfo, CommandTree } from './introspect.js';

const BUILTIN_COMMANDS = new Set(['introspect', 'version', 'skill']);

export interface SkillOptions {
  description?: string;
  whenToUse?: string;
}

function oneLine(value: string): string {
  return value.split(/\s+/).filter(Boolean).join(' ');
}

const YAML_RESERVED_START = new Set([
  '!', '&', '*', '?', '|', '>', '\'', '"', '%', '@', '`', '#', ',', '[', ']', '{', '}', '-', ':',
]);

/** Render a scalar safe for a single-line YAML block mapping value. */
function yamlScalar(value: string): string {
  if (value === '') return '""';
  const needsQuoting =
    value.includes(': ') ||
    value.includes(' #') ||
    YAML_RESERVED_START.has(value[0]) ||
    value.endsWith(':') ||
    value.trim() !== value;
  if (!needsQuoting) return value;
  const escaped = value.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
  return `"${escaped}"`;
}

function defaultDescription(name: string, userCommands: CommandInfo[]): string {
  if (userCommands.length === 0) return `Invoke the \`${name}\` CLI.`;
  const shown = userCommands.slice(0, 4).map(c => c.name);
  const suffix = userCommands.length > 4 ? '…' : '';
  return `Invoke the \`${name}\` CLI. Commands: ${shown.join(', ')}${suffix}`;
}

/** Generate a SKILL.md file from an ACLI command tree. */
export function generateSkill(
  tree: CommandTree,
  targetPath?: string,
  options: SkillOptions = {},
): string {
  const { name, version, commands } = tree;
  const userCommands = commands.filter(c => !BUILTIN_COMMANDS.has(c.name));
  const description = options.description
    ? oneLine(options.description)
    : defaultDescription(name, userCommands);

  const lines: string[] = [
    '---',
    `name: ${yamlScalar(name)}`,
    `description: ${yamlScalar(description)}`,
  ];
  if (options.whenToUse) lines.push(`when_to_use: ${yamlScalar(oneLine(options.whenToUse))}`);
  lines.push('---', '');

  lines.push(`# ${name}`, '');
  lines.push(`> Auto-generated skill file for \`${name}\` v${version}`);
  lines.push(`> Re-generate with: \`${name} skill\` or \`acli skill --bin ${name}\``);
  lines.push('');

  // Quick reference
  lines.push('## Available commands', '');
  for (const cmd of userCommands) {
    let tag = '';
    if (cmd.idempotent === true) tag = ' (idempotent)';
    else if (cmd.idempotent === 'conditional') tag = ' (conditionally idempotent)';
    lines.push(`- \`${name} ${cmd.name}\` — ${cmd.description}${tag}`);
  }
  lines.push('');

  // Detailed usage
  for (const cmd of userCommands) {
    lines.push(`## \`${name} ${cmd.name}\``, '');
    if (cmd.description) lines.push(cmd.description, '');

    if (cmd.options.length > 0) {
      lines.push('### Options', '');
      for (const opt of cmd.options) {
        const optName = opt.name.replace(/_/g, '-');
        const def = opt.default !== undefined ? ` [default: ${opt.default}]` : '';
        lines.push(`- \`--${optName}\` (${opt.type}) — ${opt.description}${def}`);
      }
      lines.push('');
    }

    if (cmd.arguments.length > 0) {
      lines.push('### Arguments', '');
      for (const arg of cmd.arguments) {
        const req = arg.required ? 'required' : 'optional';
        lines.push(`- \`${arg.name}\` (${arg.type}, ${req}) — ${arg.description}`);
      }
      lines.push('');
    }

    if (cmd.examples?.length) {
      lines.push('### Examples', '');
      for (const ex of cmd.examples) {
        lines.push('```bash', `# ${ex.description}`, ex.invocation, '```', '');
      }
    }

    if (cmd.see_also?.length) {
      const refs = cmd.see_also.map(s => `\`${name} ${s}\``).join(', ');
      lines.push(`**See also:** ${refs}`, '');
    }
  }

  // Output format
  lines.push('## Output format', '');
  lines.push(
    'All commands support `--output json|text|table`. When using `--output json`, ' +
    'responses follow a standard envelope:',
  );
  lines.push('', '```json');
  lines.push('{"ok": true, "command": "...", "data": {...}, "meta": {"duration_ms": ..., "version": "..."}}');
  lines.push('```', '');

  // Exit codes
  lines.push('## Exit codes', '');
  lines.push('| Code | Meaning | Action |');
  lines.push('|------|---------|--------|');
  lines.push('| 0 | Success | Proceed |');
  lines.push('| 2 | Invalid arguments | Correct and retry |');
  lines.push('| 3 | Not found | Check inputs |');
  lines.push('| 5 | Conflict | Resolve conflict |');
  lines.push('| 8 | Precondition failed | Fix precondition |');
  lines.push('| 9 | Dry-run completed | Review and confirm |');
  lines.push('');

  // Discovery
  lines.push('## Further discovery', '');
  lines.push(`- \`${name} --help\` — full help for any command`);
  lines.push(`- \`${name} introspect\` — machine-readable command tree (JSON)`);
  lines.push('- `.cli/README.md` — persistent reference (survives context resets)');
  lines.push('');

  const content = lines.join('\n');

  if (targetPath) {
    fs.mkdirSync(path.dirname(targetPath), { recursive: true });
    fs.writeFileSync(targetPath, content);
  }

  return content;
}

File: sdks/go/skill.go
package acli

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

var builtinCommands = map[string]struct{}{
	"introspect": {}, "version": {}, "skill": {},
}

// SkillOptions carries caller-supplied values for the SKILL.md frontmatter
// (agentskills.io).
type SkillOptions struct {
	Description string
	WhenToUse   string
}

// GenerateSkill builds SKILL.md content using default frontmatter; if path is
// non-empty, writes the file.
func GenerateSkill(tree *CommandTree, path string) (string, error) {
	return GenerateSkillWith(tree, path, SkillOptions{})
}

// GenerateSkillWith builds SKILL.md content with caller-supplied frontmatter
// options. Conforms to the agentskills.io open standard.
func GenerateSkillWith(tree *CommandTree, path string, opts SkillOptions) (string, error) {
	name := tree.Name
	ver := tree.Version

	var userCommands []CommandInfo
	for _, cmd := range tree.Commands {
		if _, skip := builtinCommands[cmd.Name]; skip {
			continue
		}
		userCommands = append(userCommands, cmd)
	}

	description := collapseWS(opts.Description)
	if description == "" {
		description = defaultSkillDescription(name, userCommands)
	}

	var b strings.Builder
	b.WriteString("---\n")
	fmt.Fprintf(&b, "name: %s\n", name)
	fmt.Fprintf(&b, "description: %s\n", description)
	if w := collapseWS(opts.WhenToUse); w != "" {
		fmt.Fprintf(&b, "when_to_use: %s\n", w)
	}
	b.WriteString("---\n\n")

	fmt.Fprintf(&b, "# %s\n\n", name)
	fmt.Fprintf(&b, "> Auto-generated skill file for `%s` v%s\n", name, ver)
	fmt.Fprintf(&b, "> Re-generate with: `%s skill` or `acli skill --bin %s`\n\n", name, name)
	b.WriteString("## Available commands\n\n")
	for _, cmd := range userCommands {
		tag := skillIdemTag(cmd)
		fmt.Fprintf(&b, "- `%s %s` — %s%s\n", name, cmd.Name, cmd.Description, tag)
	}
	b.WriteString("\n")
	for _, cmd := range userCommands {
		fmt.Fprintf(&b, "## `%s %s`\n\n", name, cmd.Name)
		if cmd.Description != "" {
			b.WriteString(cmd.Description + "\n\n")
		}
		if len(cmd.Options) > 0 {
			b.WriteString("### Options\n\n")
			for _, o := range cmd.Options {
				def := ""
				if o.Default != nil {
					def = fmt.Sprintf(" [default: %v]", o.Default)
				}
				on := strings.ReplaceAll(o.Name, "_", "-")
				fmt.Fprintf(&b, "- `--%s` (%s) — %s%s\n", on, o.Type, o.Description, def)
			}
			b.WriteString("\n")
		}
		if len(cmd.Arguments) > 0 {
			b.WriteString("### Arguments\n\n")
			for _, a := range cmd.Arguments {
				req := "optional"
				if a.Required != nil && *a.Required {
					req = "required"
				}
				fmt.Fprintf(&b, "- `%s` (%s, %s) — %s\n", a.Name, a.Type, req, a.Description)
			}
			b.WriteString("\n")
		}
		if len(cmd.Examples) > 0 {
			b.WriteString("### Examples\n\n")
			for _, ex := range cmd.Examples {
				b.WriteString("```bash\n# " + ex.Description + "\n" + ex.Invocation + "\n```\n\n")
			}
		}
		if len(cmd.SeeAlso) > 0 {
			var refs []string
			for _, s := range cmd.SeeAlso {
				refs = append(refs, fmt.Sprintf("`%s %s`", name, s))
			}
			fmt.Fprintf(&b, "**See also:** %s\n\n", strings.Join(refs, ", "))
		}
	}
	b.WriteString("## Output format\n\n")
	b.WriteString("All commands support `--output json|text|table`. When using `--output json`, " +
		"responses follow a standard envelope:\n\n```json\n" +
		`{"ok": true, "command": "...", "data": {...}, "meta": {"duration_ms": ..., "version": "..."}}` +
		"\n```\n\n")
	b.WriteString("## Exit codes\n\n")
	b.WriteString("| Code | Meaning | Action |\n|------|---------|--------|\n")
	b.WriteString("| 0 | Success | Proceed |\n| 2 | Invalid arguments | Correct and retry |\n")
	b.WriteString("| 3 | Not found | Check inputs |\n| 5 | Conflict | Resolve conflict |\n")
	b.WriteString("| 8 | Precondition failed | Fix precondition |\n| 9 | Dry-run completed | Review and confirm |\n\n")
	b.WriteString("## Further discovery\n\n")
	fmt.Fprintf(&b, "- `%s --help` — full help for any command\n", name)
	fmt.Fprintf(&b, "- `%s introspect` — machine-readable command tree (JSON)\n", name)
	b.WriteString("- `.cli/README.md` — persistent reference (survives context resets)\n\n")
	content := b.String()
	if path != "" {
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			return content, err
		}
		if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
			return content, err
		}
	}
	return content, nil
}

func collapseWS(s string) string {
	return strings.Join(strings.Fields(s), " ")
}

func defaultSkillDescription(name string, userCommands []CommandInfo) string {
	if len(userCommands) == 0 {
		return fmt.Sprintf("Invoke the `%s` CLI.", name)
	}
	n := len(userCommands)
	if n > 4 {
		n = 4
	}
	names := make([]string, n)
	for i := 0; i < n; i++ {
		names[i] = userCommands[i].Name
	}
	suffix := ""
	if len(userCommands) > 4 {
		suffix = "…"
	}
	return fmt.Sprintf("Invoke the `%s` CLI. Commands: %s%s", name, strings.Join(names, ", "), suffix)
}

func skillIdemTag(cmd CommandInfo) string {
	if cmd.Idempotent == nil {
		return ""
	}
	switch v := cmd.Idempotent.(type) {
	case bool:
		if v {
			return " (idempotent)"
		}
	case string:
		if v == "conditional" {
			return " (conditionally idempotent)"
		}
	}
	return ""
}
